src/mod/gcewing/sg/tileentity/SGPowerTE.java

```diff
@@ -78,43 +59,50 @@ public boolean hasCapability(Capability<?> capability, @Nullable EnumFacing faci
 
     @Override
     public int receiveEnergy(int maxReceive, boolean simulate) {
-        int result = storage.receiveEnergy(maxReceive, simulate);
-        energyBuffer = storage.getEnergyStored();
+        int energyReceived = Math.min((int)(energyMax - energyBuffer),maxReceive);
+        if (!simulate)
+            energyBuffer += energyReceived;
         markChanged();
-
-        return result;
+        return energyReceived;
     }
 
     @Override
     public int extractEnergy(int maxExtract, boolean simulate) {
-        int result = storage.extractEnergy(maxExtract, simulate);
-        energyBuffer = storage.getEnergyStored();
+        int energyExtracted = Math.min((int) energyBuffer, maxExtract);
+        if (!simulate)
+            energyBuffer -= energyExtracted;
         markChanged();
-
-        return result;
+        return energyExtracted;
     }
 
     @Override
     public int getEnergyStored() {
-        return storage.getEnergyStored();
+        if (energyBuffer >= Integer.MAX_VALUE) {
+            return Integer.MAX_VALUE;
+        }
+        return (int) energyBuffer;
```

Comment:
Should do `Math.floor` instead.

Reply:
I guess math.floor is the best option, I forgot that energyBuffer is Double and not Long haha
thanks